File: src/scranpy/fit_variance_trend.py
from typing import Tuple

import numpy

from . import lib_scranpy as lib


def fit_variance_trend(
    mean: numpy.ndarray,
    variance: numpy.ndarray,
    mean_filter: bool = True,
    min_mean: float = 0.1, 
    transform: bool = True, 
    span: float = 0.3,
    use_min_width: bool = False,
    min_width: float = 1,
    min_window_count: int = 200,
    num_threads: int = 1
) -> Tuple:
    """
    Fit a trend to the per-cell variances with respect to the mean.

    Args:
        mean:
            Array containing the mean (log-)expression for each gene.

        variance:
            Array containing the variance in the (log-)expression for each
            gene. This should have length equal to ``mean``.

        mean_filter:
            Whether to filter on the means before trend fitting.

        min_mean:
            The minimum mean of genes to use in trend fitting. Only used if
            ``mean_filter = True``.

        transform:
            Whether a quarter-root transformation should be applied before
            trend fitting.

        span:
            Span of the LOWESS smoother. Ignored if ``use_min_width = TRUE``.

        use_min_width:
            Whether a minimum width constraint should be applied to the LOWESS
            smoother. Useful to avoid overfitting in high-density intervals.

        min_width:
            Minimum width of the window to use when ``use_min_width = TRUE``.

        min_window_count:
            Minimum number of observations in each window. Only used if
            ``use_min_width=TRUE``.

        num_threads:
            Number of threads to use.

    Returns:
        A tuple of two arrays. The first array contains the fitted value of the
        trend for each gene while the second array contains the residual.
    """
    local_m = numpy.array(mean, dtype=numpy.float64, copy=False)
    local_v = numpy.array(variance, dtype=numpy.float64, copy=False)
    return lib.fit_variance_trend(
        local_m,
        local_v,
        mean_filter,
        min_mean,
        transform,
        span,
        use_min_width,
        min_width,
        min_window_count,
        num_threads
    )

File: lib/src/fit_variance_trend.cpp
#include <vector>
#include <stdexcept>

#include "pybind11/pybind11.h"
#include "pybind11/numpy.h"
#include "scran_variances/scran_variances.hpp"

#include "utils.h"

pybind11::tuple fit_variance_trend(
    const pybind11::array& means,
    const pybind11::array& variances,
    bool mean_filter,
    double min_mean,
    bool transform,
    double span,
    bool use_min_width,
    double min_width,
    int min_window_count,
    int num_threads)
{
    scran_variances::FitVarianceTrendOptions opt;
    opt.mean_filter = mean_filter;
    opt.minimum_mean = min_mean;
    opt.transform = transform;
    opt.span = span;
    opt.use_minimum_width = use_min_width;
    opt.minimum_width = min_width;
    opt.minimum_window_count = min_window_count;
    opt.num_threads = num_threads;

    size_t n = means.size();
    if (n != static_cast<size_t>(variances.size())) {
        throw std::runtime_error("'means' and 'variances' should have the same length");
    }

    pybind11::array_t<double> fitted(n), residuals(n);
    scran_variances::FitVarianceTrendWorkspace<double> wrk; 
    scran_variances::fit_variance_trend(
        n,
        check_numpy_array<double>(means),
        check_numpy_array<double>(variances),
        static_cast<double*>(fitted.request().ptr),
        static_cast<double*>(residuals.request().ptr),
        wrk,
        opt
    );

    pybind11::tuple output(2);
    output[0] = fitted;
    output[1] = residuals;
    return output;
}

void init_fit_variance_trend(pybind11::module& m) {
    m.def("fit_variance_trend", &fit_variance_trend);
}
